lib/prelude.dx

```diff
@@ -298,6 +298,29 @@ instance Ix Unit
 
 def iota (n:Type) [Ix n] : n=>Int = view i. ordinal i
 
+'## Fencepost index sets
+
+data Post segment:Type = UnsafeMkPost Int
+
+instance {segment} [Ix segment] Ix (Post segment)
+  get_size = \(). size segment + 1
+  ordinal = \(UnsafeMkPost i). i
+  unsafe_from_ordinal = \i. UnsafeMkPost i
+
+def left_post {n} [Ix n] (i:n) : Post n =
+  unsafe_from_ordinal (Post n) (ordinal i)
+
+def right_post {n} [Ix n] (i:n) : Post n =
+  unsafe_from_ordinal (Post n) (ordinal i + 1)
+
+interface [Ix n] NonEmpty n
+  first_ix : n
```

Comment:
Does this need to be a typeclass member? Isn't `first_ix` always equal to `unsafe_from_ordinal _ 0`?

Reply:
> Isn't first_ix always equal to unsafe_from_ordinal _ 0

Yes it is. The idea was just to require instance to supply some proof that the set is nonempty. `first_ix` seemed to serve that well, as would `last_ix` or `some_ix`. But I don't love `some_ix` because it's ill-defined.